### ISSUE TITLE
iOS Issues - RUM Events Not Population

### ISSUE THREAD

**Issue by mikhael28** [DataDog/dd-sdk-flutter#496]
**Describe what happened**
Our Flutter app uses Datadog on both platforms. Android is correctly reporting RUM Navigation events, while iOS is not.
<img width="576" alt="Screenshot 2023-10-18 at 2 12 52 PM" src="https://github.com/DataDog/dd-sdk-flutter/assets/15205259/3bc64596-b847-4f3e-9c17-dc65b9b3facc">
The above image shows Android successfully reporting navigation events, while iOS with the same config is not. 
<img width="886" alt="Screenshot 2023-10-18 at 2 13 52 PM" src="https://github.com/DataDog/dd-sdk-flutter/assets/15205259/82079706-a358-49e3-9632-20072a9703f8">

**Steps to reproduce the issue:**

```
// Dart imports:
import 'dart:developer';

// Flutter imports:
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';

// Package imports:
import 'package:datadog_flutter_plugin/datadog_flutter_plugin.dart';
import 'package:universal_platform/universal_platform.dart';

// Project imports:
import 'storage.dart';

String getRumApplicationId() {
  var rumApplicationId = '';
  if (UniversalPlatform.isAndroid) {
    rumApplicationId = 'secret';
  } else if (UniversalPlatform.isIOS) {
    rumApplicationId = 'secret';
  } else if (UniversalPlatform.isWeb) {
    rumApplicationId = 'secret';
  }
  return rumApplicationId;
}

// Rewrite log messages to remove sensitive information
LogEvent? _logEventMapper(LogEvent event) {
  print('_logEventMapper');
  if (event.message == 'overwrite me') {
    event.message = 'overwritten';
  } else if (event.message == 'stop me') {
    // Return null if you don't want a message to be sent
    return null;
  }

  return event;
}

RumViewEvent _viewEventMapper(RumViewEvent event) {
  // print('_viewEventMapper');
  if (event.view.name == 'overwrite me') {
    event.view.name = 'overwritten';
  }

  return event;
}

RumActionEvent? _actionEventMapper(RumActionEvent event) {
  print('_actionEventMapper');
  if (event.view.name == 'overwrite me') {
    event.view.name = 'overwritten';
  }

  if (event.action.target?.name == 'discard') {
    return null;
  } else if (event.action.target?.name == 'censor me!') {
    event.action.target?.name = 'xxxxxxx me';
  }

  return event;
}

RumResourceEvent? _resourceEventMapper(RumResourceEvent event) {
  print('_resourceEventMapper');
  event.resource.url = event.resource.url.replaceAll(RegExp(r'email=[^&]+'), 'email=REDACTED');

  if (event.resource.url.contains('discard')) {
    return null;
  }

  return event;
}

RumErrorEvent? _errorEventMapper(RumErrorEvent event) {
  print('_errorEventMapper');
  if (event.error.message == 'discard') {
    return null;
  }

  if (event.error.resource != null) {
    event.error.resource!.url = event.error.resource!.url.replaceAll(RegExp(r'email=[^&]+'), 'email=REDACTED');
  }

  return event;
}

RumLongTaskEvent? _longTaskEventMapper(RumLongTaskEvent event) {
  if (event.view.name == 'discard') {
    return null;
  }

  return event;
}

/// A custom [ViewInfoExtractor] that can determine the name of a route so that
/// the view can be isolated in Datadog.
///
/// The default [ViewInfoExtractor] implementation only considers [PageRoute]s.
RumViewInfo? viewInfoExtractor(Route<dynamic> route) {
  final routeName = route.settings.name;
  if (routeName != null) {
    return RumViewInfo(name: routeName);
  }

  return null;
}

final rumConfiguration = RumConfiguration(
  applicationId: getRumApplicationId(),
  detectLongTasks: true,
  reportFlutterPerformance: true,
  rumViewEventMapper: _viewEventMapper,
  rumActionEventMapper: _actionEventMapper,
  rumResourceEventMapper: _resourceEventMapper,
  rumErrorEventMapper: _errorEventMapper,
  rumLongTaskEventMapper: _longTaskEventMapper,
);

final ddSdkConfiguration = DdSdkConfiguration(
  telemetrySampleRate: 100,
  clientToken: 'secret',
  env: box.get('environment') ?? 'production',
  site: DatadogSite.us1,
  trackingConsent: TrackingConsent.granted,
  nativeCrashReportEnabled: true,
  serviceName: 'secret',
  uploadFrequency: UploadFrequency.average,
  rumConfiguration: kReleaseMode ? rumConfiguration : null,
  loggingConfiguration: LoggingConfiguration(
    sendNetworkInfo: true,
    printLogsToConsole: true,
  ),
);

Future<void> initializeDatadog() async {
  inspect(ddSdkConfiguration);
  await DatadogSdk.instance.initialize(ddSdkConfiguration);
  if (kReleaseMode) {
    await DatadogSdk.instance.rum?.initialize();
  }
}

final ddRum = DatadogSdk.instance.rum;
```

```
 Future<void> newRunApp(DdSdkConfiguration configuration, AppRunner runner) async {
    final originalOnError = FlutterError.onError;
    FlutterError.onError = (details) {
      FlutterError.presentError(details);

      DatadogSdk.instance.rum?.handleFlutterError(details);
      logger.log.e(details.exception.toString(), details.exception, details.stack);

      originalOnError?.call(details);
    };
    PlatformDispatcher.instance.onError = (error, stack) {
      DatadogSdk.instance.rum?.addErrorInfo(
        error.toString(),
        RumErrorSource.source,
        stackTrace: stack,
      );
      return false;
    };
    await DatadogSdk.instance.initialize(configuration);
    runner();
  }

  newRunApp(ddSdkConfiguration, () {
    if (UniversalPlatform.isWeb) {
        runApp(const App());
    } else {
      runApp(const App());
    }
  });
```

We use go_router, the latest version, and the only observer that is passed in is the DatadogNavigationObserver

```observers: [DatadogNavigationObserver(datadogSdk: DatadogSdk.instance)],```

In addition to this, we are receiving this message in the Flutter Debug Console

```
[VERBOSE-2:shell.cc(1004)] The 'datadog_sdk_flutter.rum' channel sent a message from native to Flutter on a non-platform thread. Platform channel messages must be sent on the platform thread. Failure to do so may result in data loss or crashes, and must be fixed in the plugin or application code creating that channel.
See https://docs.flutter.dev/platform-integration/platform-channels#channels-and-platform-threading for more information.
```

**Describe what you expected:**

**Additional context**
 - Dart/Flutter version: Flutter 3.13.5, Dart 3.1.3
 - Android/iOS OS version: iOS 17, Mac OS Sonoma, latest XCode
 - Device Model: any iOS simulator/actual model
 - Datadog SDK version: latest version on pub.dev
 - Versions of any other relevant dependencies:

**Comment by mikhael28** [comment]
Is there something that I'm missing in the setup here? If I had to guess, I would say that there is something being triggered incorrectly on the iOS side with the `await DatadogSdk.instance.rum?.initialize();` function.

**Comment by fuzzybinary** [comment]
Hi @mikhael28,

If you remove the event mappers, does iOS report properly?

I have a feeling Flutter got stricter about what threads can call into method channels. We may be incorrectly calling the mappers from off the main thread, and while I think Flutter may have always had the requirement that you should only call the method channel from the main thread, it didn't actually prevent it until recently.

If removing the mappers fixes the issue let me know and I'll get a fix out for them ASAP.

**Comment by fuzzybinary** [comment]
I just ran my integration test for the mappers and they still pass, so unfortunately that's probably not your issue (something I still intend to fix, just not what's causing this). I would remove the mappers for now anyway, unless you need them for something.

You also shouldn't need to call `DdRum.initialize` manually. It's called automatically by `DatadogSdk.initialize` if a `rumConfiguration` is present. Since this actually sets up the mapper method channel, there's a possibility that's partially causing the issue. Can you try removing it and see if that helps?

I'll add some protections to make sure double calling that issues a warning and mark the method as internal.

**Comment by mikhael28** [comment]
Hey @fuzzybinary good news - the issue was only showing up in the local dev test, even with the mappers. When deploying to prod, it reports the RUM events as expected. Well, good news / bad news I guess. All in all, things are working which is good, but the situation seems brittle.

**Comment by mikhael28** [comment]
I will be looking forward to the 2.0 release - will update this ticket at that time with anything else I've found, and hopefully be able to close this out. Updated the name of the ticket.

**Comment by fuzzybinary** [comment]
@mikhael28 That's great to hear. This may be a significant enough issue I'm going to back port it to 1.x, but I won't put it at p0 for the time being.

**Comment by fuzzybinary** [comment]
This should now be now fixed in v2.0
